### PR TITLE
BigBuf - framework / untested

### PR DESCRIPTION
**Pull request by henrygab**
# NOT TESTED

### Ready for Review:

Review of the Alpha APIv2 would be helpful.  It can be found in `mem.h`.  Feedback will define the quality of the final result.

Alpha APIv2 now exposes an API that enables up to 8k in persistent allocations, and multiple temporary allocations, while retaining the legacy 128k buffer (and its alignment, etc.).

### Info

This **_should_** just work compatibly via the old API.  Not tested yet ...this is just for early feedback.   Names (e.g., "Temporary" vs. "long-lived" allocations) are preliminary ... suggestions welcome, or we might end up with them.

Key features:
* Multiple "temporary" allocations allowed.  The purpose here is, when entering a mode, the mode may perform mode-specific allocations, which otherwise would be `static` (and thus always using RAM, even when the mode is not being used).
* When the mode exits, it is expected that the mode will free all the temporary allocations it has made.  Thus, when the mode exist, there should be zero temporary allocations remaining.
* Separate "long-lived" allocations allowed.  The purpose here is to allow a smaller set of buffers to be allocated, which are not mode-specific, but where there is value in not having the RAM always allocated.  Up to 8k is reserved for these long-lived allocations.
* Built-in API to validate that all temporary allocations have been released
* Exposure of significant (but simplified) internal state to allow reasoning about what occurs and why.

Helpful hints:
* A mode should allocate its temporary memory based on alignment requirements, requesting the allocation with the highest alignment requirement first.
* Memory buffers should be free'd in the reverse order of allocation.   Things will still work if memory is free'd in a different order, but there is currently no attempt to allocate from "holes" in the allocated memory.

<hr/><details><summary>Post v1</summary><p/>

Although not exposed yet, the new API allows multiple allocations, and re-initialization of the BigBuffer state.

This **_should_** just work compatibly via the old API.  Not tested yet ...this is just for early feedback.

Summary of changes other than `pirate/mem.c`:
* `pirate.c` -- Remove previously reserved 10k buffer
* `pirate.h` -- Remove global declaration of big buffer's size (can be API if required later)
* `pirate/memmap_buspirate5.ld` -- although not a functional change, be explicit about the need for 32k alignment of `.bss`
* `pirate/storage.c` -- Remove static assert

Summary of changes within `pirate/mem.c`:
* Big buffer is now 138k (128k + 10k) to allow for fuller allocation capability / tracking metadata.
* Primary focus is on allocation of buffers of at least 1k size (and with alignment requirements)
* Initially, just allocate large buffers from the lower side, and smaller buffers from the upper side, meeting in the middle

Later, maybe store a 128-bit map (representing 1k chunks) of what's allocated, or other tracked method to allow `free()` to be robust.

</details>

<hr/>

### CODE REVIEW

File: pirate/mem.c [hunk 1]
```diff
@@ -6,35 +6,172 @@
...
+    // finally, zero the memory before returning it.
+    memset(result, 0, totalRequestedBytes);
+    ++s_BigBufAllocationCount;
+    return result;
 }
 
+/// @brief Frees a block of memory that was allocated from the Big Buffer
+/// @param ptr Pointer to the memory block to be free'd.
+/// @note At least initially, the memory is not returned for re-use.
+/// @todo Allow the most-recent allocated memory to actually be free'd.
```
Comment (DangerousPrototypes) [inline pirate/mem.c:140]: This seems like the tricky bit (reuse). However - if this if primarily targeted at modes, then those structures all have a consistent lifecycle:
1. Enter mode
2. Allocate mode config struct
3. Exit mode
4. Release all memory

Allocating persistent ram for mode apps would have some extra complexities that I don't know how to handle. Does the app have a set of pointers that are null if unallocated, and !null when allocated? Those could also be cleared out when the mode exits.
6. Deallocate app ram from the top of the stack

Comment (henrygab) [reply pirate/mem.c:140]: If all that's desired is support for what you listed above, then the logic(+) already in this PR.

Allocation of "persistent" RAM is also possible with very small changes to the logic:

* Default / temporary allocations come from bottom (128k aligned, to start)
* Long-term allocations come from top
* BigBuffer_Reset() would reset only to the current highwatermark (leaving any long-term allocations at the top part of the address space alone)

_**Please let me know what scenarios you'd like to enable!**_

(+) The logic is enough.  Requires testing / validation / and then fixing any bugs.

Comment (henrygab) [reply pirate/mem.c]: Resolving as out-of-date with Alpha APIv2 commit

**Comment by henrygab** [comment]
Closing to avoid accidental PR.  Will generate new clean branch.

**Pull request closed without merging**